File: TERM_II/Dec_03/Longest_Palindrome.py
# Longest Palindromic Substring: return the largest palindromic (in sequence) substring from given string
# LEETCODE link: https://leetcode.com/problems/longest-palindromic-substring/description/

# Brute Force: O(n^3)
# ________________________________________________________________________________________________________________________________
def longestPalindromicSubstring_BF(s):
    def isPalindromic(substring):
        return substring == substring[::-1]
    
    n = len(s)
    res = ""
    for i in range(n):
        for j in range(i,n):
            if isPalindromic(s[i:j+1]) and len(res)<j-i+1:
                res = s[j-i+1]
    return res

# Optimised: O(n^2)
# ________________________________________________________________________________________________________________________________
def longestPalindromicSubstring_Optimised(s):
    # expands if same to both direction and returns that (palindromic) substring
    def expand(s,left,right):
        while left>=0 and right<=len(s) and s[left] == s[right]:
            left -= 1
            right += 1
        return s[left+1:right]
# left and right have moved beyond correct palindrome
# thus we need to slice from (left+1) --> (right-1) => (left+1) --> (right); (minding exclusion)

    res = ""
    n = len(s)
    for i in range(n):
        p_odd = expand(s,i,i) # checking for odd* len palindrome (i is center)
        p_even = expand(s,i,i+1) # checking for even* len palindrome

        if len(p_even) > len(res):
            res = p_even
        if len(p_odd) > len(res):
            res = p_odd
    return res
# ________________________________________________________________________________________________________________________________

File: TERM_II/Dec_03/Longest_Palindrome.java
// # Longest Palindromic Substring: return the largest palindromic (in sequence) substring from given string
// LEETCODE link: https://leetcode.com/problems/longest-palindromic-substring/description/

package College_DSA.TERM_II.Dec_03;
public class Longest_Palindrome {

    public String longestPalindrome_BF(String s) {
        int n = s.length();
        int[] pair = new int[2]; 

        for (int i=0; i<n; i++) {
            for (int j=i; j<n; j++) {
                if (checkPalindrome(s, i, j) && pair[1]-pair[0] < j-i) {
                    pair[0] = i;
                    pair[1] = j;
                }
            }
        }
        return s.substring(pair[0],pair[1]+1);
    }

    public boolean checkPalindrome(String s, int l, int r) {
        while (l < r) {
            if (s.charAt(l) != s.charAt(r)) return false;
            l++;
            r--;
        }
        return true;
    }

    public String longestPalindrome(String s) {
        int n = s.length();
        int[] pair = new int[2]; 
        int l, r;

        for (int i=0; i<n; i++) {
            int[] oddRange = expand(i,i,s);
            l = oddRange[0];
            r = oddRange[1];
            
            if (pair[1]-pair[0]+1 < r-l+1) {
                pair[0] = l;
                pair[1] = r;
            }
            
            int[] evenRange = expand(i,i+1,s);
            l = evenRange[0];
            r = evenRange[1];

            if (pair[1]-pair[0]+1 < r-l+1) {
                pair[0] = l;
                pair[1] = r;
            }
            
        }   
        return s.substring(pair[0],pair[1]+1);
    }

    public int[] expand(int l, int r, String s) {
        while (l>=0 && r<s.length() && s.charAt(l) == s.charAt(r)) {
            l--;
            r++;
        }
        return new int[]{l+1,r-1};
    }
}